torch/_dynamo/guards.py

```diff
@@ -611,6 +642,145 @@ def _produce_guard_code(
+        ast.Pow,
+        ast.LShift,
+        ast.RShift,
+        ast.BitOr,
+        ast.BitXor,
+        ast.BitAnd,
+        ast.FloorDiv,
+        # Unary Operation nodes
+        ast.Invert,
+        ast.Not,
+        ast.UAdd,
+        ast.USub,
+        # Compare Operation nodes
+        ast.Eq,
+        ast.NotEq,
+        ast.Lt,
+        ast.LtE,
+        ast.Gt,
+        ast.GtE,
+        ast.Is,
+        ast.IsNot,
+        ast.In,
+        ast.NotIn,
+    )
+
+    class CSEVisitor(ast.NodeVisitor):
+        IGNORE = object()
+
+        def __init__(self) -> None:
+            self._expr_counter: Dict[str, int] = collections.defaultdict(lambda: 0)
+
+        def visit(self, node: ast.AST) -> Any:
+            if _ast_unparse_implemented(node) and not isinstance(
+                node, PyExprCSEPass.IGNORED_NODE_TYPES
+            ):
+                self._expr_counter[_ast_unparse(node)] += 1
+            super().visit(node)
+
+    class CSETransformer(ast.NodeTransformer):
+        def __init__(
+            self,
+            expr_counter: Dict[str, int],
+            preface: List[str],
+            gen_name: Callable[[], str],
+        ) -> None:
+            super().__init__()
+            self._expr_counter = expr_counter
+            self._preface = preface
+            self._gen_name = gen_name
+            self._expr_to_name: Dict[str, str] = {}
+
+        def visit(self, node: ast.AST) -> Any:
+            if _ast_unparse_implemented(node) and not isinstance(
+                node, PyExprCSEPass.IGNORED_NODE_TYPES
+            ):
+                expr = _ast_unparse(node)
+
+                # Replacement only occurs if a given expression is used more
+                # than once.
+                if self._expr_counter[expr] > 1:
+                    if expr not in self._expr_to_name:
+                        # Parent 'visit' is called so that we CSE the inner expressions first.
+                        #
+                        # The resulting expression is used as right-hand-side of the variable
+                        # assignment. i.e. we are CSE-ing the children before the parents.
+                        #
+                        # Indexing still uses the old 'node', since that's what was counted
+                        # by the 'NodeVisitor'.
+                        node_ = super().visit(node)
+                        expr_ = _ast_unparse(node_)
+                        var_name = self._gen_name()
+                        self._preface.append(f"{var_name} = {expr_}")
+                        self._expr_to_name[expr] = var_name
+                    else:
+                        var_name = self._expr_to_name[expr]
+                    return ast.Name(var_name, ast.Load())
+
+            return super().visit(node)
+
+    def __init__(self) -> None:
+        self._counter = 0
+
+    def _new_var(self, prefix: str = "_var") -> str:
+        name = f"{prefix}{self._counter}"
+        self._counter += 1
```

Comment:
When /why do we need to increment? do all new vars just get a number assigned?

Reply:
This is assigning names to each CSE'd expr, so obvs they need different names

Reply:
> So, I think I can make this PR better (I think I can finish the implementation tomorrow) by:
> 
> 1. We can make it work without `ast.unparse` (eliminating the quadratic time) by using the column information from parsing
> 2. Use `ALLOW_NODE_TYPES` (the inverse of what we have now) so as to filter the nodes interesting to the CSE, namely: `Subscript`, `Attribute`, and `Call` (for `getattr` calls)
> 3. Divide the guard function into non-ShapeEnv guards and ShapeEnv guards (which will free us from the try-catch)

We already divide ordering, shape_env guards get added separately. It would not be a stretch to do this.

Reply:
Right, but I was wondering if we can use the prior unique names, there should be no collisions.

Reply:
The first version of CSE (in the other PR) did exactly that. The problems were:

- We still needed the counter, since the expr-var mapping wasn't injective (i.e. collisions could happen)
- Increased compilation time
- Ad-hoc logic for replacing prohibited symbols

Reply:
> (FYI: I will be off until May 1)

If need be, I can help finish this PR and land it under your attribution.